primitives/npos-elections/src/phragmen.rs

```diff
@@ -0,0 +1,201 @@
+/// assignments, `assignment.distribution.map(|p| p.deconstruct()).sum()` fails to fit inside
+/// `UpperOf<P>`. A user of this crate may statically assert that this can never happen and safely
+/// `expect` this to return `Ok`.
+pub fn seq_phragmen<AccountId: IdentifierT, P: PerThing>(
+	rounds: usize,
+	initial_candidates: Vec<AccountId>,
+	initial_voters: Vec<(AccountId, VoteWeight, Vec<AccountId>)>,
+	balance: Option<(usize, ExtendedBalance)>,
+) -> Result<ElectionResult<AccountId, P>, &'static str> where ExtendedBalance: From<InnerOf<P>> {
+	let (candidates, voters) = setup_inputs(initial_candidates, initial_voters);
+
+	let (candidates, mut voters) = seq_phragmen_core::<AccountId>(
+		rounds,
+		candidates,
+		voters,
+	)?;
+
+	if let Some((iterations, tolerance)) = balance {
+		// NOTE: might create zer-edges, but we will strip them again when we convert voter into
+		// assignment.
+		let _iters = balancing::balance::<AccountId>(&mut voters, iterations, tolerance);
+	}
+
+	let mut winners = candidates
+		.into_iter()
+		.filter(|c_ptr| c_ptr.borrow().elected)
+		.take(rounds) // defensive only: seq-phragmen-core returns only up to rounds.
+		.collect::<Vec<_>>();
+
+	// sort winners based on desirability.
+	winners.sort_unstable_by_key(|c_ptr| c_ptr.borrow().round);
+
+	let mut assignments = voters.into_iter().filter_map(|v| v.into_assignment()).collect::<Vec<_>>();
+	let _ = assignments.iter_mut().map(|a| a.try_normalize()).collect::<Result<(), _>>()?;
+	let winners = winners.into_iter().map(|w_ptr|
+		(w_ptr.borrow().who.clone(), w_ptr.borrow().backed_stake)
+	).collect();
+
+	Ok(ElectionResult { winners, assignments })
+}
+
+/// Core implementation of seq-phragmen.
+///
+/// This is the internal implementation that works with the types defined in this crate. see
+/// `seq_phragmen` for more information. This function is left public in case a crate needs to use
+/// the implementation in a custom way.
+///
+/// To create th inputs needed for this function, see [`crate::setup_inputs`].
+///
+/// This can only fail if the normalization fails.
+pub fn seq_phragmen_core<AccountId: IdentifierT>(
+	rounds: usize,
+	candidates: Vec<CandidatePtr<AccountId>>,
+	mut voters: Vec<Voter<AccountId>>,
+) -> Result<(Vec<CandidatePtr<AccountId>>, Vec<Voter<AccountId>>), &'static str> {
+	// we have already checked that we have more candidates than minimum_candidate_count.
+	let to_elect = rounds.min(candidates.len());
+
+	// main election loop
+	for round in 0..to_elect {
+		// loop 1: initialize score
+		for c_ptr in &candidates {
+			let mut candidate = c_ptr.borrow_mut();
+			if !candidate.elected {
+				// 1 / approval_stake == (DEN / approval_stake) / DEN. If approval_stake is zero,
+				// then the ratio should be as large as possible, essentially `infinity`.
+				if candidate.approval_stake.is_zero() {
+					candidate.score = Rational128::from_unchecked(DEN, 0);
```

Comment:
The purpose of this is to set the score here to the largest possible value. This is because later on we look for the min score, so this will have the least chance. Based on this, I will implement a proper `Bounded::max_value` for Rational128 and use it here.